internal/dataplane/kongstate/kongstate.go

```diff
@@ -690,3 +683,303 @@ func maybeLogKongIngressDeprecationError(logger logr.Logger, services []*corev1.
+		collection := ks.CustomEntities[entity.Spec.EntityType]
+		collection.Entities = append(collection.Entities, CustomEntity{
+			Object:              parsedEntity,
+			k8sKongCustomEntity: entity,
+		})
+	}
+
+	ks.sortCustomEntities()
+}
+
+// fetchEntitySchema fetches schema of an entity by its type and stores the schema in its custom entity collection
+// as a cache to avoid excessive calling of Kong admin APIs.
+func (ks *KongState) fetchEntitySchema(schemaGetter SchemaGetter, entityType string) (EntitySchema, error) {
+	collection, ok := ks.CustomEntities[entityType]
+	if ok {
+		return collection.Schema, nil
+	}
+	// Use `context.Background()` here because `BuildKongConfig` does not provide a context.
+	schema, err := schemaGetter.Get(context.Background(), entityType)
+	if err != nil {
+		return EntitySchema{}, err
+	}
+	return ExtractEntityFieldDefinitions(schema), nil
+}
+
+// fillCustomEntityForeignFields fills the "foreign" fields of a custom Kong entity
+// if it refers to a service/route/consumer.
+// Because Kong gateway requires the "foreign" fields to use IDs of the referred entity,
+// it fills ID of the referred entity and fill the ID of the entity to the field.
+// So the function has the side effect that the referred entity will have a generated fixed ID.
+// It does not support fields referring to other types of entities.
+func (ks *KongState) fillCustomEntityForeignFields(
+	logger logr.Logger,
+	k8sEntity *kongv1alpha1.KongCustomEntity,
+	schema EntitySchema,
+	parsedEntity map[string]any,
+	pluginRelEntities PluginRelatedEntitiesRefs,
+	workspace string,
+) {
+	logger = logger.WithValues("entity_namespace", k8sEntity.Namespace, "entity_name", k8sEntity.Name)
+	// Find referred entity via the plugin in its spec.parentRef.
+	// Then we can fetch the referred service/route/consumer from the reference relations of the plugin.
+	parentRef := k8sEntity.Spec.ParentRef
+	var namespace string
+	// Abort if the parentRef is empty or does not refer to a plugin.
+	if parentRef == nil ||
+		(parentRef.Group == nil || *parentRef.Group != kongv1alpha1.GroupVersion.Group) {
+		return
+	}
+	if parentRef.Kind == nil || (*parentRef.Kind != "KongPlugin" && *parentRef.Kind != "KongClusterPlugin") {
+		return
+	}
+	// Extract the plugin key to get the plugin relations.
+	if parentRef.Namespace == nil || *parentRef.Namespace == "" {
+		namespace = k8sEntity.Namespace
+	} else {
+		namespace = *parentRef.Namespace
+	}
+	pluginKey := namespace + ":" + parentRef.Name
+	// Get the relations with other entities of the plugin.
+	rels, ok := pluginRelEntities.RelatedEntities[pluginKey]
+	if !ok {
+		return
+	}
+	logger.V(util.DebugLevel).Info("fetch references via plugin", "plugin_key", pluginKey)
+
+	// Traverse through the fields of the entity and fill the "foreign" fields with IDs of referring entities.
+	// Note: this procedure will make referred services'/routes'/consumers' ID to be filled. Should we set `FillIDs` feature gate as a prerequisite?
```

Comment:
We definitely do not want to make this implicit. If that's required let's enforce this at the CLI flag level.

Reply:
Since we have decided to put this behind a feature gate, do you think we should force it to be enabled only when `FillIDs`  feature gate and print an error log if custom entities is enabled with `FillIDs` disabled?

Reply:
> Since we have decided to put this behind a feature gate

Did we? I thought we're leaving this as is with the CLI flag disabled.

> ... do you think we should force it to be enabled only when FillIDs feature gate and print an error log if custom entities is enabled with FillIDs disabled?

If (my understanding might be incomplete on the dependency chain here) fillIDs feature flag is required for this to work then yes, I'd expect it to be required and KIC should report an error and exit if the required `FillIDs` FF is not enabled.

Reply:
Added comment here to tell that the `FillIDs` feature gate is required.